teps/0004-tekton-integrations-annotations-statuses.md

```diff
@@ -0,0 +1,689 @@
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Integration Annotations
+
+[Annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/)
+of the format `<name>.integrations.tekton.dev/<value>` to will be used to
+annotate PipelineRun and TaskRun objects with integration metadata. This will
+not be strictly enforced, in order to allow interoperability with other tools /
+annotation namespaces, but we should recommend this as a default by convention.
+
+This fits into the annotations model - from the Kubernetes docs:
+
+> Fields managed by a declarative configuration layer. Attaching these fields as
+> annotations distinguishes them from default values set by clients or servers,
+> and from auto-generated fields and fields set by auto-sizing or auto-scaling
+> systems.
+>
+> Build, release, or image information like timestamps, release IDs, git branch,
+> PR numbers, image hashes, and registry address.
+>
+> User or tool/system provenance information, such as URLs of related objects
+> from other ecosystem components.
+
+This allows for integrations to annotate arbitrary data without needing to
+modify the Pipeline execution spec, with the expectation that another reconciler
+/ tool will notice and act on these annotations.
+
+**Examples:**
+
+- GitHub
+  ```
+  github.integrations.tekton.dev/owner: wlynch
+  github.integrations.tekton.dev/repo: test
+  github.integrations.tekton.dev/sha: 4d2f673c600a0f08c3f6737bc3f5c6a6a15d62e6
+  ```
+- Webhook
+  ```
+  webhook.integrations.tekton.dev/url: https://example.com:8080
+  ```
+- Slack
+  ```
+  slack.integrations.tekton.dev/url: https://tektoncd.slack.com
+  slack.integrations.tekton.dev/channel: build-cop
+  ```
+
+### Integration Status
+
+Key questions we will need to answer is - did an integration complete
+successfully for my pipeline? if not, why?
+
+In Tekton's
+[dogfooding cluster](https://github.com/tektoncd/plumbing/blob/master/docs/dogfooding.md)
+today, this answer is somewhat obfuscated. When a pipeline is complete, the
+cluster kicks off a cloud event containing information about the PR to update.
+This is picked up by another Trigger, which then updates the PR. It is difficult
+to tell from the initial PipelineRun whether a PR was updated, and if there was
+a problem, why.
+
+To solve this, we should add additional status fields in
+TaskRunStatus/PipelineRunStatus, specifically for integration data. This should
+include information such as:
+
+- What integration the status is for (e.g. `github.integrations.tekton.dev`).
+- What object generation the status is for.
+- What was the outcome of the integration for this object (e.g. success,
+  failure).
+- (optional) More details of status outcome.
```

Comment:
One thing to keep in mind is that `TaskRun` and `PipelineRun` are more or less "immutable", so I am not sure how fully useful the object generation will be as… except from the `cancel` field in spec, the rest of the spec won't change.

Reply:
For TaskRun and PipelineRun `spec`, I 100% agree. `status` is where I see room for flexibility (since these are mutated during Pipeline execution anyway).

Re object/observed generation: I'm not particularly tied to this. I'm mostly borrowing this from our existing [TaskRunStatus](https://pkg.go.dev/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1?tab=doc#TaskRunStatus) type which uses the [Knative Status duck type](https://pkg.go.dev/knative.dev/pkg/apis/duck/v1beta1?tab=doc#Status).